File: threadsflow/modules/content.py
"""Thread publishing, direct replies to any post, quoting, liking, and tree ingestion."""
from __future__ import annotations
import json
import time
from typing import Dict, Any, List, Optional
from ..types import ThreadPost, DiscussionTree

class ContentModule:
    """Operations on Thread posts, replies, and discussion trees."""

    def __init__(self, client):
        self._client = client

    def publish_text(self, text: str, reply_control: str = "everyone") -> dict:
        """
        Publish a new top-level text thread.
        reply_control options: 'everyone', 'accounts_you_follow', 'mentioned_only'.
        """
        data = {
            "caption": text,
            "text_post_app_info": json.dumps({"reply_control": reply_control}),
        }
        return self._client._post("/api/v1/media/configure_text_post_app_feed/", data=data)

    def reply(self, parent_post_id: str | int, text: str, reply_control: str = "everyone") -> dict:
        """
        Reply directly to ANY public thread post without Meta App Review restrictions.
        """
        data = {
            "caption": text,
            "text_post_app_info": json.dumps({
                "reply_to_post_id": str(parent_post_id),
                "reply_control": reply_control,
            }),
        }
        return self._client._post("/api/v1/media/configure_text_post_app_feed/", data=data)

    def quote(self, quoted_post_id: str | int, text: str) -> dict:
        """Quote an existing thread post with added commentary."""
        data = {
            "caption": text,
            "text_post_app_info": json.dumps({
                "quoted_post_id": str(quoted_post_id),
            }),
        }
        return self._client._post("/api/v1/media/configure_text_post_app_feed/", data=data)

    def get_thread(self, post_id: str | int) -> DiscussionTree:
        """
        Retrieve complete nested discussion hierarchy (root post, parent, and recursive replies).
        """
        res = self._client._get(f"/api/v1/text_post_app/{post_id}/text_post_app_thread/")
        threads_data = res.get("containing_thread", res)
        root = ThreadPost.from_dict(threads_data.get("thread_items", [{}])[0])
        replies = []
        for item in res.get("reply_threads", []):
            for t_item in item.get("thread_items", []):
                replies.append(ThreadPost.from_dict(t_item))
        return DiscussionTree(root_post=root, replies=replies)

    def thread_likers(self, post_id: str | int) -> List[dict]:
        """Fetch list of accounts that liked this thread."""
        res = self._client._get(f"/api/v1/media/{post_id}/likers/")
        return res.get("users", [])

    def like(self, post_id: str | int) -> dict:
        """Like a thread post."""
        return self._client._post(f"/api/v1/media/{post_id}/like/")

    def unlike(self, post_id: str | int) -> dict:
        """Unlike a thread post."""
        return self._client._post(f"/api/v1/media/{post_id}/unlike/")

    def repost(self, post_id: str | int) -> dict:
        """Repost / amplify a thread to your followers."""
        return self._client._post(f"/api/v1/media/{post_id}/repost/")

    def unrepost(self, post_id: str | int) -> dict:
        """Remove a previously reposted thread."""
        return self._client._post(f"/api/v1/media/{post_id}/unrepost/")

    def delete(self, post_id: str | int) -> dict:
        """Delete an owned thread post."""
        data = {"media_id": str(post_id)}
        return self._client._post(f"/api/v1/media/{post_id}/delete/", data=data)

File: ts/src/modules/content.ts
import { DiscussionTree, parseThreadPost } from "../types";

export interface IThreadsHttpClient {
  get(path: string, params?: Record<string, any>): Promise<any>;
  post(path: string, data?: Record<string, any>): Promise<any>;
}

export class ContentModule {
  private client: IThreadsHttpClient;

  constructor(client: IThreadsHttpClient) {
    this.client = client;
  }

  /**
   * Publish a new top-level text thread.
   * @param text Post content
   * @param replyControl 'everyone' | 'accounts_you_follow' | 'mentioned_only'
   */
  public async publishText(text: string, replyControl: string = "everyone"): Promise<any> {
    const data = {
      caption: text,
      text_post_app_info: JSON.stringify({ reply_control: replyControl }),
    };
    return this.client.post("/api/v1/media/configure_text_post_app_feed/", data);
  }

  /**
   * Reply directly to ANY public thread post without Meta App Review restrictions.
   * @param parentPostId Thread post ID
   * @param text Reply text
   * @param replyControl 'everyone' | 'accounts_you_follow' | 'mentioned_only'
   */
  public async reply(
    parentPostId: string | number,
    text: string,
    replyControl: string = "everyone",
  ): Promise<any> {
    const data = {
      caption: text,
      text_post_app_info: JSON.stringify({
        reply_to_post_id: String(parentPostId),
        reply_control: replyControl,
      }),
    };
    return this.client.post("/api/v1/media/configure_text_post_app_feed/", data);
  }

  /**
   * Quote an existing thread post with added commentary.
   */
  public async quote(quotedPostId: string | number, text: string): Promise<any> {
    const data = {
      caption: text,
      text_post_app_info: JSON.stringify({
        quoted_post_id: String(quotedPostId),
      }),
    };
    return this.client.post("/api/v1/media/configure_text_post_app_feed/", data);
  }

  /**
   * Retrieve complete nested discussion hierarchy (root post, parent, and recursive replies).
   */
  public async getThread(postId: string | number): Promise<DiscussionTree> {
    const res = await this.client.get(`/api/v1/text_post_app/${postId}/text_post_app_thread/`);
    const threadsData = res.containing_thread || res;
    const threadItems = threadsData.thread_items || [{}];
    const root = parseThreadPost(threadItems[0] || {});

    const replies = [];
    for (const item of res.reply_threads || []) {
      for (const tItem of item.thread_items || []) {
        replies.push(parseThreadPost(tItem));
      }
    }

    return {
      rootPost: root,
      replies,
      hasMore: Boolean(res.has_more),
      cursor: res.cursor,
    };
  }

  /**
   * Fetch list of accounts that liked this thread.
   */
  public async threadLikers(postId: string | number): Promise<any[]> {
    const res = await this.client.get(`/api/v1/media/${postId}/likers/`);
    return res.users || [];
  }

  /**
   * Like a thread post.
   */
  public async like(postId: string | number): Promise<any> {
    return this.client.post(`/api/v1/media/${postId}/like/`);
  }

  /**
   * Unlike a thread post.
   */
  public async unlike(postId: string | number): Promise<any> {
    return this.client.post(`/api/v1/media/${postId}/unlike/`);
  }

  /**
   * Repost / amplify a thread to your followers.
   */
  public async repost(postId: string | number): Promise<any> {
    return this.client.post(`/api/v1/media/${postId}/repost/`);
  }

  /**
   * Remove a previously reposted thread.
   */
  public async unrepost(postId: string | number): Promise<any> {
    return this.client.post(`/api/v1/media/${postId}/unrepost/`);
  }

  /**
   * Delete an owned thread post.
   */
  public async delete(postId: string | number): Promise<any> {
    const data = { media_id: String(postId) };
    return this.client.post(`/api/v1/media/${postId}/delete/`, data);
  }
}
